### DISCUSSION TITLE
Split Discord Module

### DISCUSSION THREAD

**Discussion by Cahllagerfeld** [EddieHubCommunity/api#238] [Ideas]
See the discussion [here](https://github.com/EddieHubCommunity/api/discussions/224). We may want to split the current `Discord` to `Profiles` and `DiscordEvents`

The `Profile` one could have the following DTO:
```ts
export interface SocialItem {
  name: string;
  url: string;
}

export interface Milestone {
    title: string;
    date: string;
    icon: string;
    color: string;
    description: string;
    url: string;
  }
 
export interface Profiles {
  _id: string; //userid from Discord in case Discord is the single-source of truth for this, when using the discord-id we may populate this even with the stats from the discord module
  bio: string;
  githubUsername: string;
  avatarURL: string;
  profiles: {
    twitter: SocialItem;
    github: SocialItem; // the ones added here need to be discussed, but imo definitely not generic
  };
  milestones: Milestone[]
}
```

The `DiscordEvents` the following:
```ts
export interface DiscordStats {
  _id: string; //uderid from Discord
  eventName: string;
}
```
when using the same ID, populating each other shouldn't be hard so we can even get the Data from both objects.

Especially for `Profile` we definitely need to think of a solution for creating or deleting a milestone then, if its stored in an array. Maybe this needs to be extracted to a collection as well 🤔 

Let me know what you think and if its worth to create a bit more in-detail concepts for this

**Comment by suitupstan** [comment]
The interface SocialItem should have a type for icon determination:

```ts
export type supportedSocials = 'discord' | 'twitter' | 'github' | 'other'

export interface SocialItem {
  name: string;
  url: string;
  type: supportedSocials;
}
```

The `url` in Milestone, isn't required, also a better way of dates, like TimeStamps to provide cross-timezone time would be great

```ts
export interface Milestone {
    title: string;
    date: Date;
    icon: string;
    color: string;
    description: string;
    url?: string;
  }
```

Also, profiles, we need the avatar key if we use Discord ids and not github ones

```ts
export interface Profiles {
  _id: string; //userid from Discord in case Discord is the single-source of truth for this, when using the discord-id we may populate this even with the stats from the discord module
  bio: string;
  profiles: {
    twitter: SocialItem;
    github: SocialItem; // the ones added here need to be discussed, but imo definitely not generic
  };
  milestones: Milestone[]
}
```

I think there is lot more than just DiscordStats for DiscordEvents 🤔 There was a recent discussion on making Katra signups work with our events on Discord

**Reply by Cahllagerfeld** [reply]
>
>The interface SocialItem should have a type for icon determination:
>
I think we can get this from the prop-name of  `obj.profiles.twitter` for example

>
>The url in Milestone, isn't required, also a better way of dates, like TimeStamps to provide cross-timezone time would be great
>
I'd definitely store the timezones as `ISO-String` in the database, timezone conversion needs to be done in the Frontend, nothing the api needs to care about

>
>Also, profiles, we need the avatar key if we use Discord ids and not github ones
>
Thats a good point 👍

**Reply by suitupstan** [reply]
I think a we should have a icon value, instead of just a few configured prop-names because it is not necessary that I only want to use the social icons, I use  the book icon for my blog link and so on...

**Reply by Cahllagerfeld** [reply]
Then we have the wild-west data again that I wanted to prevent. What happens if someone adds a icon which isnt available, do we really want to allow scenarios where someone adds a twitter-profile, but with discord icon for example? Imo we need to be consistent with this.

**Reply by suitupstan** [reply]
If we wanna do something like this, We would need to configure a lot of these, just to be able to migrate.

**Reply by Cahllagerfeld** [reply]
I'm not sure if we migrate the existing ones tbh

**Comment by Cahllagerfeld** [comment]
When creating the module, would you split Socials and Milestones to seperate Entities, or would you just put them as properties from the socials-entity @eddiejaoude ? I'm still not quite sure if its the best to transfer them as `regular` properties, or instead of own entities. When having them on their own we can add or remove them easier I think, but I'm not sure if it makes thing just more complicated for now. Please tell me what you think about this

**Reply by eddiejaoude** [reply]
Sorry for my slow reply, I think this has been done now?
